Too many open files

Workflow as follows:

Read in the file
`temp_path = ‘pathpathpath.qptiff’`

Create a Slide Object
`mif_slide = SlideData(temp_path, backend = “bioformats”, slide_type = types.Vectra)`

I am creating tiles like so

Create tiles from the slide
`tiles = mif_slide.generate_tiles(shape = 256)`
`list_of_tiles = list(tiles)`

However, I get the error:
```

---------------------------------------------------------------------------
JavaException                             Traceback (most recent call last)
~/.local/lib/python3.8/site-packages/bioformats/formatreader.py in init_reader(self)

~/.local/lib/python3.8/site-packages/bioformats/formatreader.py in setId(self, path)

~/.local/lib/python3.8/site-packages/javabridge/jutil.py in call(o, method_name, sig, *args)

~/.local/lib/python3.8/site-packages/javabridge/jutil.py in fn(*args)

JavaException: /mnt/disks/ip_tiles/immunoprofile-raw/immunoprofile/Lung_Cancer/IP-18-J00055/ROI_ANALYSIS/IP_18_J00055_Scan1.qptiff (Too many open files)

During handling of the above exception, another exception occurred:

AttributeError                            Traceback (most recent call last)
/tmp/ipykernel_7604/3079534046.py in <module>
      1 #Prototype the pipeline on a set of tiles
----> 2 list_of_tiles = list(tiles)
      3 # next(tiles)

~/pathml/pathml/core/slide_data.py in generate_tiles(self, shape, stride, pad, **kwargs)
    379             pathml.core.tile.Tile: Extracted Tile object
    380         """
--> 381         for tile in self.slide.generate_tiles(shape, stride, pad, **kwargs):
    382             # add masks for tile, if possible
    383             # i.e. if the SlideData has a Masks object, and the tile has coordinates

~/pathml/pathml/core/slide_backends.py in generate_tiles(self, shape, stride, pad, level)

~/pathml/pathml/core/slide_backends.py in extract_region(self, location, size, level)

~/.local/lib/python3.8/site-packages/bioformats/formatreader.py in __init__(self, path, url, perform_init)

~/.local/lib/python3.8/site-packages/bioformats/formatreader.py in init_reader(self)

AttributeError: 'JavaException' object has no attribute 'message'
```

I wrote a test and fixed this by putting ImageReader in a with scope.
When I ```javabridge.kill_vm()``` we get the error:
```
E           RuntimeError: Failed to start Java VM

../../anaconda3/envs/pathml/lib/python3.8/site-packages/javabridge/jutil.py:319: RuntimeError
```
Apparently this is the expected behavior https://stackoverflow.com/questions/51554355/python-javabridge-virtual-machine-only-works-once and we should not stop the VM.

--- a/pathml/core/slide_backends.py
+++ b/pathml/core/slide_backends.py
@@ -25,11 +25,13 @@ try:
     import javabridge
     from bioformats.metadatatools import createOMEXMLMetadata
 except ImportError:
-    raise Exception("""Installation of PathML not complete. Please install openjdk8, bioformats, and javabridge:
+    raise Exception(
+        """Installation of PathML not complete. Please install openjdk8, bioformats, and javabridge:
             conda install openjdk==8.0.152
             pip install javabridge==1.0.19 python-bioformats==4.0.0
-            
-            For detailed installation instructions, please see https://github.com/Dana-Farber-AIOS/pathml/""")
+
+            For detailed installation instructions, please see https://github.com/Dana-Farber-AIOS/pathml/"""
+    )
 
 
 class SlideBackend:
@@ -314,27 +316,45 @@ class BioFormatsBackend(SlideBackend):
                 f"input size {size} invalid. Must be a tuple of integer coordinates of len<2"
             )
         javabridge.start_vm(class_path=bioformats.JARS, max_heap_size="100G")
-        reader = bioformats.ImageReader(str(self.filename), perform_init=True)
-        # expand size
-        size = list(size)
-        arrayshape = list(size)
-        for i in range(len(self.shape)):
-            if i > len(size) - 1:
-                arrayshape.append(self.shape[i])
-        arrayshape = tuple(arrayshape)
-        array = np.empty(arrayshape)
-        sample = reader.read(
-            z=0, t=0, rescale=False, XYWH=(location[0], location[1], size[0], size[1])
-        )
-        # if series is set to read only one channel, explicitly read c
-        if len(sample.shape) == 2:
-            for z in range(self.shape[2]):
-                for c in range(self.shape[3]):
+        with bioformats.ImageReader(str(self.filename), perform_init=True) as reader:
+            # expand size
+            size = list(size)
+            arrayshape = list(size)
+            for i in range(len(self.shape)):
+                if i > len(size) - 1:
+                    arrayshape.append(self.shape[i])
+            arrayshape = tuple(arrayshape)
+            array = np.empty(arrayshape)
+            sample = reader.read(
+                z=0,
+                t=0,
+                rescale=False,
+                XYWH=(location[0], location[1], size[0], size[1]),
+            )
+            # if series is set to read only one channel, explicitly read c
+            if len(sample.shape) == 2:
+                for z in range(self.shape[2]):
+                    for c in range(self.shape[3]):
+                        for t in range(self.shape[4]):
+                            slicearray = reader.read(
+                                z=z,
+                                t=t,
+                                series=c,
+                                rescale=False,
+                                XYWH=(location[0], location[1], size[0], size[1]),
+                            )
+                            slicearray = np.asarray(slicearray)
+                            # some file formats read x, y out of order, transpose
+                            if slicearray.shape[:2] != array.shape[:2]:
+                                slicearray = np.transpose(slicearray)
+                            array[:, :, z, c, t] = slicearray
+            # if series is set to read all channels, read all c simultaneously
+            elif len(sample.shape) == 3:
+                for z in range(self.shape[2]):
                     for t in range(self.shape[4]):
                         slicearray = reader.read(
                             z=z,
                             t=t,
-                            series=c,
                             rescale=False,
                             XYWH=(location[0], location[1], size[0], size[1]),
                         )
@@ -342,25 +362,10 @@ class BioFormatsBackend(SlideBackend):
                         # some file formats read x, y out of order, transpose
                         if slicearray.shape[:2] != array.shape[:2]:
                             slicearray = np.transpose(slicearray)
-                        array[:, :, z, c, t] = slicearray
-        # if series is set to read all channels, read all c simultaneously
-        elif len(sample.shape) == 3:
-            for z in range(self.shape[2]):
-                for t in range(self.shape[4]):
-                    slicearray = reader.read(
-                        z=z,
-                        t=t,
-                        rescale=False,
-                        XYWH=(location[0], location[1], size[0], size[1]),
-                    )
-                    slicearray = np.asarray(slicearray)
-                    # some file formats read x, y out of order, transpose
-                    if slicearray.shape[:2] != array.shape[:2]:
-                        slicearray = np.transpose(slicearray)
-                        slicearray = np.moveaxis(slicearray, 0, -1)
-                    array[:, :, z, :, t] = slicearray
-        else:
-            raise Exception("image format not supported")
+                            slicearray = np.moveaxis(slicearray, 0, -1)
+                        array[:, :, z, :, t] = slicearray
+            else:
+                raise Exception("image format not supported")
         array = array.astype(np.uint8)
         return array
 

--- a/tests/core_tests/conftest.py
+++ b/tests/core_tests/conftest.py
@@ -7,7 +7,7 @@ import anndata
 import numpy as np
 # define fixtures here, and use them throughout the other tests in core_tests/
 import pytest
-from pathml.core import SlideData, SlideDataset, Tiles
+from pathml.core import SlideData, SlideDataset, Tile, VectraSlide, types
 
 
 @pytest.fixture
@@ -87,3 +87,10 @@ def slide_dataset_with_tiles(tile, example_slide_data_with_tiles):
     ]
     slide_dataset = SlideDataset(slide_list)
     return slide_dataset
+
+
+@pytest.fixture
+def vectra_slide():
+    temp_path = "tests/testdata/small_vectra.qptiff"
+    vectra_slide = VectraSlide(temp_path, backend="bioformats", slide_type=types.Vectra)
+    return vectra_slide

--- a/tests/core_tests/test_slide_backends.py
+++ b/tests/core_tests/test_slide_backends.py
@@ -142,3 +142,8 @@ def test_dicom_coords_index_conversion():
     for index, coords in check.items():
         assert backend._index_to_coords(index) == coords
         assert backend._coords_to_index(coords) == index
+
+
+def test_bioformats_vm_handling(vectra_slide):
+    tiles = vectra_slide.generate_tiles(shape = 10)
+    list_of_tiles = list(tiles)

